src/pmdas/gpfs/pmdagpfs.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2011 SGI.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMDAGPFS 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmdagpfs\f1 \- gpfs filesystem statistics performance metrics domain agent (PMDA)
.SH DESCRIPTION
\f3pmdagpfs\f1 is a Performance Metrics Domain Agent (\s-1PMDA\s0) which
exports metric values from the
.I /usr/lpp/mmfs/bin/mmpmon
program to provide information on mounted gpfs filesystems.
.SH INSTALLATION
Install the gpfs PMDA by using the Install script as root:
.sp 1
.RS +4
.ft B
.nf
# cd $PCP_PMDAS_DIR/gpfs
# ./Install
.fi
.ft P
.RE
.sp 1
.PP
To uninstall, do the following as root:
.sp 1
.RS +4
.ft B
.nf
# cd $PCP_PMDAS_DIR/gpfs
# ./Remove
.fi
.ft P
.RE
.sp 1
\fBpmdagpfs\fR is launched by \fIpmcd\fR(1) and should never be executed
directly. The Install and Remove scripts notify \fIpmcd\fR(1) when the
agent is installed or removed.
.SH FILES
.IP "\fB$PCP_PMDAS_DIR/gpfs/Install\fR" 4
installation script for the \fBpmdagpfs\fR agent
.IP "\fB$PCP_PMDAS_DIR/gpfs/Remove\fR" 4
undo installation script for the \fBpmdagpfs\fR agent
.IP "\fB$PCP_LOG_DIR/pmcd/gpfs.log\fR" 4
default log file for error messages from \fBpmdagpfs\fR
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fR are used to parameterize
the file and directory names used by \fBPCP\fR. On each installation, the
file \fB/etc/pcp.conf\fR contains the local values for these variables.
The \fB$PCP_CONF\fR variable may be used to specify an alternative
configuration file, as described in \fIpcp.conf\fR(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmcd (1)
and
.BR mmpmon (8).

.\" control lines for scripts/man-spell
.\" +ok+ pmdagpfs gpfs
.\" +ok+ mmpmon mmfs lpp {all from /usr/lpp/mmfs/bin/mmpmon}


CHANGELOG
pcp-6.2.2 (29 May 2024)
    - Work-in-progress, not yet released.

pcp-6.2.1 (10 April 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdauwsgi: (new) instrumentation from uWSGI servers
	pmdalinux: new hugepage metrics from sysfs (fixed size hugepages)
	pmdalinux: new filesys.uuid and filesys.type metrics
	pmdalinux: new TcpExt metrics from /proc/net/netstat
	pmdalinux: new softnet metrics from newer kernel versions
	pmdalinux: new /proc/vmstat khugepaged metrics
	pmdabpf: updates to add eBPF networking metrics for pcp-atop
	pmdaopenmetrics: fix script error diagnostic cascading exception
    - Client tools and utilities:
	pcp2openmetrics: (new) push PCP metrics in OpenMetrics format
	pmcheck: (new) interrogate and control PCP components
	pmcheck: Redis server detection and agent recommendation
	pmcheck: PostgreSQL server detection and agent recommendation
	pmcheck: uwSGI server detection and agent recommendation
	pcp-atopsar: fix TZ mishandling relating to day-crossover
	pcp-dstat: fix an I/O config typo, sum reads and writes
	pcp-htop: sync latest htop-3.3.0 fixes
	pmrep.conf: use bpf instead of bcc pmda for proc net metrics
	pmrep.conf: proc-essential cosmetic improvements
	pmrep.conf: remove vmeff field from sar -B config
	pmrepconf: use pmrep-internal representations for each metric
    - Server-side utilities and log management scripts:
	pmlogger: bizarre timezone fix (github #1936)
	pmlogger: improve handling of a change in metadata semantics
	pmlogdump: add support for multi-archive contexts
	pmproxy: disable Redis protocol proxying by default
	runaspcp: remove hard-coded "pcp" for user and group
	init scripts: drop conditonal use of setpriv(1) or runuser(1)
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix logmeta.c SEGV with V3 archives
	libpcp: fix memleak on repeated __pmFixPMNSHashTab calls
	libpcp: add pmgetopt pmflush() call for warnings
	libpcp: fixes for V3 archives and multi-archive contexts
	libpcp: export __pmLogChangeArchive() interface to tools
	libpcp: fix for PMNS with multi-archive contexts
	libpcp: small derived metrics bug fix (github #1921)
    - Build, infrastructure and packaging updates:
	Linux MX distro support (Debian without systemd)
	riscv: configure pmdabpf_arch value correctly
	tar packages: fix perl modules install in postinstall script
	eBPF updates: add submodule for blazesym, sync to latest
	debian build: add postrm for pcp-zeroconf
	build: removed cppcheck from 'make check', too flakey
	container: Fedora 40 container testing added
	constainer: fixed Fedora base image location (quay.io)
	build: add support for riscv64 to pcp rpm spec files
    - Security Enhanced Linux:
	selinux policy: small tweak for pmie signal handling
    - Documentation:
	man pages: document pmlogger_check and pmlogger_daily SaveLogs
	man pages: man-spell and corrections
	docs: updated donations page with new fiscal sponsor details

pcp-6.2.0 (12 February 2024)
    - Client tools and utilities:
	pmlogger: default to creating version 3 PCP archives
	pmie, pmlogger, pmcd, pmproxy: init script hardening to
	    run less as root:root, more as pcp:pcp and more safely
	pcp-reboot-init: new systemd service for $PCP_RC_DIR setup
	pcp2arrow: new, export metrics to parquet format for pandas
	pcp-atop: deterministic process sort when handling NULLs
	pcp-atop: update to v2.10.0 upstream code
	pcp-htop: update to latest upstream fixes
	pcp-geolocate: function with older versions of python
	pcp-meminfo: resolve possible 'broken pipe' diagnostics
	pcp-ss: fix handling of listening state sockets to match ss
	pcp-ss: fix the UDP UNCONN vs LISTEN state handling
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: improvements for process accounting metrics
	pmdalinux: add pgpromote* and pgdemote* memory metrics
	pmdalinux: add metric sysfs.module.zswap.max_pool_percent
	pmdalinux: add new SNMP metric network.ip.outtransmits
	pmdasummary: send state change 0 (no change) back to pmcd
	pmdapostgresql: support the psycopg3 python module as well
	pmdaopenmetrics: kepler support - live metrics and logging
    - Server-side utilities and log management scripts:
	pmlogrewrite, pmlogredact: fix for missing data volumes
	pmcd: cosmetic change to avoid pipe2 name clash
	runaspcp: new helper utility for unprivileged rc scripts
	pmnsmerge: be more defensive in qsort comparison routine
	pmlogsize: fix qsort comparison function handling of equality
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: pmOpenLog refactor for handling previous log files
	libpcp: new __pmCleanMapDir routine to aid log culling
	libpcp & pmcd: READY-NOT-READY hardening
	libpcp: avoid possible thread issues (coverity)
	libpcp: refactor pmDupContext for thread safety
	libpcp_import: add guard in pmiPutMark to prevent SIGSEGV
    - Build, infrastructure and packaging updates:
	build: changes to re-enable builds for FreeBSD 12, 13, 14
	build: add diffutils rpm dep as pmdaproc script uses diff
	build: add main.yml for building and publishing container
	build: resolve compiler warnings in various tools/libraries
	packaging: use tmpfiles.d method to create PCP_RUN_DIR
	build: packaging changes for openSUSE 15.5 and 15.6
	debian: do not ship empty directory /usr/lib/pkgconfig
	build: el10 systemd-presets for pmcd/pmie/pmlogger enabling
	build: drop rpm dep on bpftool as local version used only
	build: change perl YAML::XS::LibYAML module to YAML::XS
    - Security Enhanced Linux:
	selinux: policy improvements for CentOS/RHEL 7 and 8
    - Documentation:
	docs: Fedora 39 found some tbl(1)/troff(1) issues, fix
	docs: add an explicit code-of-conduct for PCP
	docs: pmns(5) note that metric names must only contain ASCII
	docs: update the BPF README to describe bpftool reality

pcp-6.1.1 (17 November 2023)
    - Client tools and utilities:
	pmie: add -o/--format to control output format with archives

src/pmdas/gpfs/pmdagpfs.pl
#
# Copyright (c) 2015 Martins Innus.  All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#

use strict;
use warnings;
use PCP::PMDA;

our $noval = PM_ERR_APPVERSION;

use vars qw( $pmda );

# FS Indom.  Will use cluster:fs as instance
our  $gpfs_fs_indom = 0;

our $MMPMON_CMD = "echo \"fs_io_s\" | /usr/lpp/mmfs/bin/mmpmon -s -p";

# Configuration files for overriding the location of MMPMON_CMD, etc, mostly for testing purposes
for my $file (pmda_config('PCP_PMDAS_DIR') . '/gpfs/gpfs.conf', 'gpfs.conf') {
        eval `cat $file` unless ! -f $file;
}

# Check env variable for MMPMON_CMD to use.  Should be a cmd.  Easiest to use "echo test_file" for QA
if ( defined $ENV{"GPFS_MMPMON_CMD"} ) {
	$MMPMON_CMD = $ENV{"GPFS_MMPMON_CMD"};
}

# The stats hash is keyed on cluster:filesystem
our %h_gpfs = ();

# Parse the client io stats
# Looks like this:
#
# _fs_io_s_ _n_ 10.10.1.1 _nn_ myhost _rc_ 0 _t_ 1421869154 _tu_ 766776 _cl_ gpfs.foobar.com _fs_ gpfs0 _d_ 107 _br_ 3592793024 _bw_ 2552776479 _oc_ 6758 _cc_ 6542 _rdc_ 2923 _wc_ 176495 _dir_ 1367 _iu_ 1418
#
# which is defined like:
#
# mmpmon node 10.10.1.1 name myhost fs_io_s OK
# cluster:        gpfs.foobar.com
# filesystem:     gpfs0
# disks:                 107
# timestamp:      1421869103/978437
# bytes read:     3592793024
# bytes written:  2552776479
# opens:                6758
# closes:               6542
# reads:                2923
# writes:             176495
# readdir:              1367
# inode updates:        1418

sub gpfs_get_io_stats{
	open(MMPMON, "$MMPMON_CMD |") || die "Can't open pipe: $MMPMON_CMD\n";
	while (<MMPMON>){
		my $line = $_;

		# Split on whitesapce and hope they only tack onto the end if they add fields
		my @stats = split(/\s+/,$line);

		my $cluster = $stats[12];
		my $fs = $stats[14];
		my $disks = $stats[16];
		my $bread = $stats[18];
		my $bwrite = $stats[20];
		my $opens = $stats[22];
		my $closes = $stats[24];
		my $reads = $stats[26];
		my $writes = $stats[28];
		my $readdir = $stats[30];
		my $iupdates = $stats[32];

		my $id = "$cluster:$fs";

		$h_gpfs{$id} = {}; # {} is an anonymous hash ref
		$h_gpfs{$id}->{'gpfs.fsios.cluster'} = $cluster;
		$h_gpfs{$id}->{'gpfs.fsios.filesystem'} = $fs;
		$h_gpfs{$id}->{'gpfs.fsios.disks'} = $disks;
		$h_gpfs{$id}->{'gpfs.fsios.read_bytes'} = $bread;
		$h_gpfs{$id}->{'gpfs.fsios.write_bytes'} = $bwrite;
		$h_gpfs{$id}->{'gpfs.fsios.opens'} = $opens;
		$h_gpfs{$id}->{'gpfs.fsios.closes'} = $closes;
		$h_gpfs{$id}->{'gpfs.fsios.reads'} = $reads;
		$h_gpfs{$id}->{'gpfs.fsios.writes'} = $writes;
		$h_gpfs{$id}->{'gpfs.fsios.readdir'} = $readdir;
		$h_gpfs{$id}->{'gpfs.fsios.inode_updates'} = $iupdates;
	}
	close( MMPMON );
}

#
# fetch is called once by pcp for each refresh and then the fetch callback is
# called to query each statistic individually
#
sub gpfs_fetch {
	gpfs_get_io_stats();

	our $pmda->replace_indom($gpfs_fs_indom, \%h_gpfs);
}

sub gpfs_fetch_callback {
	my ($cluster, $item, $inst) = @_; 

	if( $cluster == 0 ){
		# gpfs fs_io_s client stats
		my $lookup = pmda_inst_lookup($gpfs_fs_indom, $inst);
		return (PM_ERR_INST, 0) unless defined($lookup);

		my $pmid_name = pmda_pmid_name($cluster, $item)
			or die "Unknown metric name: cluster $cluster item $item\n";

		return ($lookup->{$pmid_name}, 1);
	}
	else{
		return (PM_ERR_PMID, 0);
	}
}

# the PCP::PMDA->new line is parsed by the check_domain rule of the PMDA build
# process, so there are special requirements:  no comments, the domain has to
# be a bare number.
#
our $pmda = PCP::PMDA->new('gpfs', 135);

# Metrics

#  gpfs fs_io_s client stats - cluster 0

# fs_io_s

$pmda->add_metric(pmda_pmid(0,1), PM_TYPE_STRING, $gpfs_fs_indom,
		PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
		'gpfs.fsios.cluster',
		'Cluster name',
		'');

$pmda->add_metric(pmda_pmid(0,2), PM_TYPE_STRING, $gpfs_fs_indom,
		PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
		'gpfs.fsios.filesystem',
		'Filesystem name',
		'');

$pmda->add_metric(pmda_pmid(0, 3), PM_TYPE_U64, $gpfs_fs_indom,
		PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
		"gpfs.fsios.disks",
		'Number of disks in this filesystem',
		'');

$pmda->add_metric(pmda_pmid(0, 4), PM_TYPE_U64, $gpfs_fs_indom,
		PM_SEM_COUNTER, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
		"gpfs.fsios.read_bytes",
		'Read bytes counter',
		'');

$pmda->add_metric(pmda_pmid(0, 5), PM_TYPE_U64, $gpfs_fs_indom,
		PM_SEM_COUNTER, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
		"gpfs.fsios.write_bytes",
		'Write bytes counter',
		'');

$pmda->add_metric(pmda_pmid(0, 6), PM_TYPE_U64, $gpfs_fs_indom,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		"gpfs.fsios.opens",
		'Number of open calls',
		'');

$pmda->add_metric(pmda_pmid(0, 7), PM_TYPE_U64, $gpfs_fs_indom,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		"gpfs.fsios.closes",
		'Number of close calls',
		'');

$pmda->add_metric(pmda_pmid(0, 8), PM_TYPE_U64, $gpfs_fs_indom,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		"gpfs.fsios.reads",
		'Number of read calls',
		'');

$pmda->add_metric(pmda_pmid(0, 9), PM_TYPE_U64, $gpfs_fs_indom,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		"gpfs.fsios.writes",
		'Number of write calls',
		'');

$pmda->add_metric(pmda_pmid(0, 10), PM_TYPE_U64, $gpfs_fs_indom,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		"gpfs.fsios.readdir",
		'Number of readdir calls',
		'');

$pmda->add_metric(pmda_pmid(0, 11), PM_TYPE_U64, $gpfs_fs_indom,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		"gpfs.fsios.inode_updates",
		'Number of inode update calls',
		'');

&gpfs_get_io_stats;

$gpfs_fs_indom = $pmda->add_indom($gpfs_fs_indom, {}, '', '');

$pmda->set_fetch(\&gpfs_fetch);
$pmda->set_fetch_callback(\&gpfs_fetch_callback);

$pmda->run;
